go/consensus/cometbft/apps/roothash/state/state.go

```diff
@@ -303,6 +376,82 @@ func (s *MutableState) SetRuntimeState(ctx context.Context, state *roothash.Runt
 	if err := s.ms.Insert(ctx, ioRootKeyFmt.Encode(&state.Runtime.ID), ioRoot); err != nil {
 		return api.UnavailableStateError(err)
 	}
+
+	// Clean previously stored state and I/O roots if we're over the maximum.
+	params, err := s.ConsensusParameters(ctx)
+	if err != nil {
+		return api.UnavailableStateError(err)
+	}
+	maxStored := params.MaxPastRootsStored
+
+	// Add state and I/O roots for this round if enabled.
+	if maxStored > 0 {
+		newRound := state.CurrentBlock.Header.Round
+		newRoots := cbor.Marshal(roothash.PastRoots{
+			StateRoot: state.CurrentBlock.Header.StateRoot,
+			IORoot:    state.CurrentBlock.Header.IORoot,
+		})
+
+		// Delete the oldest root to make room for the new one.
+		// We ignore the error because the root might not be there, so it's OK
+		// if this fails.
+		_ = s.ms.Remove(ctx, pastRootsFmt.Encode(&state.Runtime.ID, newRound-uint64(maxStored)))
+
+		if err := s.ms.Insert(ctx, pastRootsFmt.Encode(&state.Runtime.ID, newRound), newRoots); err != nil {
+			return api.UnavailableStateError(err)
+		}
+	}
+
+	return nil
+}
+
+// ShrinkPastRoots deletes extra past stored roots for all runtimes that have
+// over the given number of stored roots.
+// This is used when reducing the MaxPastRootsStored consensus parameter in
+// changeParameters in go/consensus/cometbft/apps/roothash/messages.go.
+func (s *MutableState) ShrinkPastRoots(ctx context.Context, max uint32) error {
+	// Go through all runtimes, so we can delete extra stored past
+	// roots for each one, where it's needed.
+	runtimes, err := s.Runtimes(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, r := range runtimes {
+		id := r.Runtime.ID
+		numStoredRoots := s.PastRootsCount(ctx, id)
+		if numStoredRoots <= max {
+			// Nothing to delete.
+			continue
+		}
+
+		numPastRootsToDelete := numStoredRoots - max
+
+		it := s.is.NewIterator(ctx)
+		defer it.Close()
+
+		var keysToRemove [][]byte
```

Comment:
I don't feel comfortable skipping the verification steps, especially since they're not going to be the bottleneck here.
We could save a tiny bit of RAM by doing removals on the fly, though, so that's maybe worth doing as an optimization later if necessary.
I've modelled this after `RemoveExpiredEvidence`, which also first collects all the keys and then removes them.